neps/nep-0413.md

```diff
@@ -0,0 +1,138 @@
+  callbackUrl?: string; // Optional, applicable to browser wallets (e.g. MyNearWallet). The URL to call after the signing process. Defaults to `window.location.href`.
+}
+```
+
+### Structure
+`signMessage` must embed the input `message`, `receiver` and `nonce` into the following predefined structure:
+
+```rust
+struct Payload {
+  message: string; // The same message passed in `SignMessageParams.message`
+  receiver: string; // The same receiver passed in `SignMessageParams.receiver`
+  nonce: [u8; 32]; // The same nonce passed in `SignMessageParams.nonce`
+}
+```
+
+### Signature
+In order to create a signature, `signMessage` must:
+1. Convert the `Payload` into its [JSON JCS](https://www.rfc-editor.org/rfc/rfc8785) string representation (i.e. a JSON string, with alphabetically ordered attributes).
+2. Prepend the `NEP0413:` string to the result from step 1.
+3. Compute the `SHA256` hash of the result from step 2.
+4. Sign the resulting `SHA256` hash from step 3 using a **full-access** key.
+
+> If the wallet does not hold any `full-access` keys, then it must return an error.
+
+### Example
+Assuming that the `signMessage` method was invoked, and that:
+- The input `message` is `"hi"`
+- The input `receiver` is `"myapp.com"`
+- The input `nonce` is `[0,...,31]`
+- The wallet stores a full-access private key
+
+The wallet must construct and sign the following `SHA256` hash:
+
+```jsx
+sha256.hash(`NEP0413:` + `{"message":"hi","receiver":"myapp.com","nonce":"[0,...,31]"}`)
+```
+
+### Output Interface
+`signMessage` must return an object containing the **base64** representation of the `signature`, and all the data necessary to verify such signature. 
+
+```jsx
+interface SignedMessage {
+  accountId: string; // The account name to which the publicKey corresponds as plain text (e.g. "alice.near")
+  publicKey: string; // The public counterpart of the key used to sign, expressed as a string with format "<key-type>:<base-64-key-bytes>"
+  signature: string; // The base64 representation of the signature.
+}
+```
+
+### Returning the signature
+#### Web Wallets
+Web Wallets, such as [MyNearWallet](https://mynearwallet.com), should directly return the `SignedMessage` to the `SignMessageParams.callbackUrl`, passing both `accountId` and `publicKey` as strings, and the `signature` as an URL fragment. This is: `<callbackUrl>?accountId=<accountId>&publicKey=<publicKey>#signature=<signature>`.
+
+If the signing process fails, then the wallet must return an error message as a string parameter: `<callbackUrl>?error=<error-message-string>`.
+
+#### Other Wallets
+Non-web Wallets, such as [Ledger](https://www.ledger.com) can directly return the `SignMessageOut` (in preference as a JSON object) and raise an error on failure.
+
+## References
+A full example on how to implement the `signMessage` method can be [found here](https://github.com/gagdiez/near-login/blob/main/tests/authentication/auth.ava.ts#L27-#L65).
+
+## Drawbacks
```

Comment:
Currently, NEAR ledger app will be unable to sign these messages without an update to make it aware of this format -- as it is currently written it only works with pure transactions

Reply:
Added it as a drawback.